ddtrace/provider.py

```diff
@@ -50,17 +55,23 @@ def _has_active_context(self):
         :returns: Whether we have an active context
         :rtype: bool
         """
-        return self._local._has_active_context()
+        ctx = _DD_CONTEXTVAR.get()
+        return ctx is not None
 
-    def activate(self, context):
+    def activate(self, ctx):
         """Makes the given ``context`` active, so that the provider calls
         the thread-local storage implementation.
         """
-        return self._local.set(context)
+        _DD_CONTEXTVAR.set(ctx)
 
     def active(self):
         """Returns the current active ``Context`` for this tracer. Returned
         ``Context`` must be thread-safe or thread-local for this specific
         implementation.
         """
-        return self._local.get()
+        ctx = _DD_CONTEXTVAR.get()
+        if not ctx:
+            ctx = Context()
+            self.activate(ctx)
+
+        return ctx
```

Comment:
This does not look thread-safe, but I think this is not a problem right since the context is per-thread per definition right?

Reply:
if it is fine, probably means we need a good comment/description here to explain the expectations (test are good to reinforce when possible)

Reply:
There is no change in functionality here.

Thread-safety is handled by underlying contextvar which is per thread/executor. It's possible that if there were some kind of async environment with multithreaded tasks to have races when setting the context because the variable will be stored in the task and shared amongst the threads. I don't think this is worth dealing with and the `get()...activate()` race is no longer a thing starting in #1936 (the safety should be ensured by `contextvars`).

Reply:
You're right, I missed this was just code moving. It's not worse so let's go.
Sorry about the noise.

Reply:
no problem at all, was worth the thought exercise IMO 🙂